File: nhance/nhance/report/purchase_order_workflow_actions/purchase_order_workflow_actions.js
// Copyright (c) 2016, Epoch and contributors
// For license information, please see license.txt
/* eslint-disable */

frappe.query_reports["Purchase Order Workflow Actions"] = {
	"filters": [
		{
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "reqd": 1,
            "default": frappe.datetime.add_months(frappe.datetime.month_start(), -1),
            "width": "80"
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "reqd": 1,
            "default": frappe.datetime.add_months(frappe.datetime.month_end(), -1)
        },
	{
	    "fieldname": "status",
            "label": __("Status"),
            "fieldtype": "Link",
            "options": "Workflow State"
	}
	],
	onload: function(report) {
        //console.log("Make PO and Transfer Existing/Required Quantity..........");
        report.page.add_inner_button(__("Workflow Action"),
            function() {
		
                var reporter = frappe.query_reports["Purchase Order Workflow Actions"];
                reporter.make_PO_and_transfer_qty(report);
            });
    },
    make_PO_and_transfer_qty: function(report) {
	  var start_date = frappe.query_report.get_filter_value("from_date");
	   var end_date = frappe.query_report.get_filter_value("to_date");
	   var status = frappe.query_report.get_filter_value("status");
	   var purchase_order = "";
	   var purchase_data = []
	   if (status == ""){
	   	var purchase_order = get_purchase_order(start_date,end_date);
		}
	  if (status != ""){
		var start = start_date
		var with_status_details = {
					"start_date":start_date,
					"end_date":end_date,
					"status":status
				}
		purchase_data.push(with_status_details);
		var purchase_order = purchase_order_with_status(purchase_data);
		}
		//console.log("purchase order ----------"+JSON.stringify(purchase_order));
		var fields = [];
		var data1 = {
			    "fieldtype":"Read Only",
                            "label": "Purchase Order",
                            "fieldname": "purchase_order"

                        }
		 fields.push(data1);
		var data2=	{
			    
                            "fieldtype": "Column Break",
                            "fieldname": "column_breaks"
			}
		 fields.push(data2);
		var data3	= {   
			    "fieldtype":"Read Only",
                            "label": "Status",
                            "fieldname": "status"
			}
		 fields.push(data3);
		var data4=	{
                            "fieldtype": "Column Break",
                            "fieldname": "column_break2s"
			}
		 fields.push(data4);
		var data5=	{
			    "fieldtype":"Read Only",
                            "label": "Posting Date",
                            "fieldname": "posting_date"
			}
		 fields.push(data5);
		var data6=	{
			   "fieldtype": "Section Break",
                            "fieldname": "section_breaks"
			}		
			
                fields.push(data6);
		//console.log("fields-----------------"+fields);
		 for(var i =0; i < purchase_order.length;i++){
		//console.log("hello");
		
		var data = {
                            "fieldtype": "Check",
                            "label": purchase_order[i].name,
                            "fieldname": purchase_order[i].name

                        }
			fields.push(data);
		
		var data11 =	{
                            "fieldtype": "Column Break",
                            "fieldname": "column_break" + i + ""
			}
			fields.push(data11);
			
		var data12 =	{
			    "fieldtype":"Read Only",
                            "label": purchase_order[i].workflow_state,
                            "fieldname": purchase_order[i].workflow_state
			}
		fields.push(data12);
		
		var data13 ={
                            "fieldtype": "Column Break",
                            "fieldname": "column_break2" + i + ""
			}
		fields.push(data13);
		
		var data14 ={
			   "fieldtype":"Read Only",
                            "label": purchase_order[i].creation,
                            "fieldname": purchase_order[i].creation
			}
			fields.push(data14);
		
		var data15 ={
			     
			   "fieldtype": "Section Break",
                            "fieldname": "section_break"+i+""
			}
		fields.push(data15);
		}
		  var dialog = new frappe.ui.Dialog({
		 'fields': fields,
	    primary_action: function(){
		dialog.hide();
		var dialog_data = dialog.get_values();
		var purchase_name = [];
		  Object.keys(dialog_data).forEach(function(key) {
		   if (dialog_data[key] == 1) {
			console.log("purchase order---------------"+key);
			for (var pur =0; pur<purchase_order.length; pur++){
				if(purchase_order[pur].name == key){
					purchase_details = {
							"name": purchase_order[pur].name,
							"status":purchase_order[pur].workflow_state
						}
					purchase_name.push(purchase_details)
					
				
				}
			
			}
				}
			});
			//console.log("purchase name -------------"+purchase_name);
			//var actions = frappe.workflow.get_all_transition_actions("Purchase Order");
			
			if (!frappe.model.has_workflow("Purchase Order")) 
				return frappe.msgprint("worklow is not active");
			
			if (purchase_name.length > 0){
			var next_state = [];
			var role = "";
			var workflow_data = get_workflow_details(purchase_name)
			//console.log("workflow_data--------------"+JSON.stringify(workflow_data));
			for (var wrk =0; wrk < workflow_data.length; wrk++){
				var status = workflow_data[wrk].action;
				 role = workflow_data[wrk].allowed;
				next_state.push(status);
				//console.log("status--------------"+status);
			}
			var check_role = get_roles(frappe.session.user, role);
			console.log("next stage-----------"+next_state);
			console.log("check_role-----------"+check_role);
			if(check_role == role){
			//if(frappe.user_roles.indexOf(role) == 1){
			var s_dialog =new frappe.ui.Dialog({
				//title "please select  workflow Action";
				"fields":[
					{
					"fieldtype": "Select",
                           		 "fieldname": "actions",
					 "label":"Actions",
				         "options": next_state	
					}],
				  primary_action: function(){
					s_dialog.hide();
					var action_dialog = s_dialog.get_values();
					console.log("action_dialog------------"+JSON.stringify(action_dialog));
					var action = action_dialog.actions;
					console.log("action --------------"+action);
					var success = set_workflow_purchase_order(action,purchase_name);
					console.log("success------------",success);
					if(success == 1){
						frappe.msgprint("workflow has updated ");
					}
				}
			 });
			s_dialog.show();
			}else{
				frappe.msgprint('"'+frappe.session.user+'"'+" Don't have permission "+'"'+role+'"' +" to perform "+'"'+next_state+'"');
			}
		}
		else{
			frappe.msgprint("Please select atleast one purchase order");
		}
		
	    }
	});
	//dialog.fields_dict.ht.$wrapper.html('Hello World');
	/*for(var j = 0; j < purchase_order.length; j++){
	 var name = purchase_order[j].name;
	console.log("name---------"+name);
	
	  dialog.fields_dict.name.input.onclick = function() {
		console.log("checked");
		}
	}*/
	dialog.show();
	//console.log("fields-----------------------"+JSON.stringify(fields));
}	
}
function get_purchase_order(start_date,end_date){
	 var purchase = "";
    frappe.call({
        method: "nhance.nhance.report.purchase_order_workflow_actions.purchase_order_workflow_actions.purchase_order_detials",
        args: {
            "start_date": start_date,
           "end_date": end_date
        },
        async: false,
        callback: function(r) {
            //console.log("ItemValuationRateFromStockBalance::" + JSON.stringify(r.message));
            purchase = r.message;

        } //end of call-back function..
    }); //end of frappe call..
    return purchase;

}

function get_workflow_details(purchase_name){
	 var worflow = "";
    frappe.call({
        method: "nhance.nhance.report.purchase_order_workflow_actions.purchase_order_workflow_actions.workflow_details",
        args: {
            "purchase_name": purchase_name
        },
        async: false,
        callback: function(r) {
         
            worflow = r.message;

        } //end of call-back function..
    }); //end of frappe call..
    return worflow;

}

function purchase_order_with_status(purchase_data){
	
	 var purchase = "";
    frappe.call({
        method: "nhance.nhance.report.purchase_order_workflow_actions.purchase_order_workflow_actions.purchase_order_with_status",
        args: {
            "purchase_data": purchase_data
        },
        async: false,
        callback: function(r) {
      
            purchase = r.message;

        } 
    }); 
    return purchase;


}

function get_roles(user, role){
	var purchase = "";
    frappe.call({
        method: "nhance.nhance.report.purchase_order_workflow_actions.purchase_order_workflow_actions.user_details",
        args: {
            "user": user,
	    "role":role
        },
        async: false,
        callback: function(r) {
	if (r.message.length > 0){
      
            purchase = r.message[0].role;
	}
        } 
    }); 
    return purchase;


}

function set_workflow_purchase_order(action,purchase_name){
	var purchase = "";
    frappe.call({
        method: "nhance.nhance.report.purchase_order_workflow_actions.purchase_order_workflow_actions.ready_to_update_workflow_state",
        args: {
            "action": action,
	    "purchase_name":purchase_name
        },
        async: false,
        callback: function(r) {
	if (r.message){
      
            purchase = r.message;
	}
        } 
    }); 
	console.log("purchase-------------flag"+purchase);
    return purchase;


}

File: nhance/nhance/report/purchase_order_workflow_actions/purchase_order_workflow_actions.py
# Copyright (c) 2013, Epoch and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from frappe import _, msgprint
from datetime import date,datetime
def execute(filters=None):
	start_date = filters.get("from_date")
	end_data = filters.get("to_date")
	status = filters.get("status")
	columns, data = [], []
	columns = get_columns()

	purchase_order = get_purchase_order(start_date,end_data,status)
	for purchase in purchase_order:
		posting_date = purchase.creation
		posting_date = posting_date.strftime('%d-%m-%Y')
		schedule_date = purchase.schedule_date
		schedule_date = schedule_date.strftime('%d-%m-%Y')
		items_details = get_items_detials(purchase.name)
		tax_details = get_tax_detials(purchase.name)
		payment = get_payment_details(purchase.name)
		tax_amount = 0.0
		for tax in tax_details:
			tax_amount += tax.tax_amount
			tax_amount = round(tax_amount,2)
		data.append([posting_date,purchase.name,purchase.workflow_state,items_details[0].project,purchase.supplier,items_details[0].item_code,items_details[0].stock_qty,
		items_details[0].stock_uom,items_details[0].qty,items_details[0].uom,
		items_details[0].net_rate,items_details[0].net_amount,tax_amount,purchase.grand_total,payment[0].payment_term])
		total_net_amount = 0.0
		total_net_amount += items_details[0].net_amount
		for items in items_details:
				if items_details[0].item_code != items.item_code:
					data.append(["","","","","",items.item_code,items.stock_qty,items.stock_uom,items.qty,items.uom,items.net_rate,items.net_amount,"",""])
					total_net_amount += items.net_amount
		data.append(["Total","","","","","","","","","","",total_net_amount,tax_amount,purchase.grand_total])
	return columns, data


def get_columns():
	"""return columns"""
	columns = [
		_("Posting Date")+"::120",
		_("Purchase Order")+":Link/Purchase Order:120",
		_("Status")+"::120",
		_("Project")+"::120",
		_("Supplier")+"::120",
		_("Item Code")+"::120",
		_("Stock Qty")+"::120",
		_("Stock UOM")+"::120",
		_("Purchase Qty")+"::120",
		_("Purchase UOM")+"::120",
		_("Net Purchase Rate")+"::120",
		_("Total Net Value")+"::120",
		_("Total Tax")+"::120",
		_("Grand Total")+"::100",
		_("Payment Terms")+"::120"
		 ]
	return columns

def get_purchase_order(start_date,end_data,status):
	start_date = str(start_date)+" 00:00:00:00"
	end_data = str(end_data)+" 23:59:59"

	purchase = frappe.db.sql("""select * from `tabPurchase Order` where creation >= %s and creation <= %s and workflow_state = %s""",(start_date,end_data,status), as_dict =1)

	return purchase

def get_purchase_order_with_status(start_date,end_data,status):
	start_date = str(start_date)+" 00:00:00:00"
	end_data = str(end_data)+" 23:59:59"

	purchase = frappe.db.sql("""select * from `tabPurchase Order` where creation >= %s and creation <= %s and workflow_state = %s""",(start_date,end_data,status), as_dict =1)

	return purchase

@frappe.whitelist()
def purchase_order_detials(start_date,end_date):
	purchase = get_purchase_order(start_date,end_date)
	return purchase

@frappe.whitelist()
def purchase_order_with_status(purchase_data):
	#with_status_details = eval(with_status_details)
	start_date = ""
	end_date = ""
	status = ""
	for status_with in eval(purchase_data):

		start_date = status_with['start_date']
		end_date = status_with['end_date']
		status = status_with['status']

	purchase = get_purchase_order_with_status(start_date,end_date,status)
	return purchase
@frappe.whitelist()
def workflow_details(purchase_name):
	status = ""
	for purchase in eval(purchase_name):
		status = purchase['status']
	workflow = frappe.db.sql("""select * from `tabWorkflow Transition` where state = %s""",status, as_dict =1)
	#role =  frappe.get_roles(frappe.session.user)
	#print "role-----------",role

	#print "workflow--------------",workflow
	return workflow

@frappe.whitelist()
def user_details(user,role):
	user_data = frappe.db.sql("""select role from `tabHas Role` where parent = '"""+user+"""' AND role =%s """,role,as_dict=1)
	if user_data is not None:
		return user_data
	else:
		return None

@frappe.whitelist()
def ready_to_update_workflow_state(action,purchase_name):
	flag = 0
	now = datetime.now()

	#print "now =", now
	# dd/mm/YY H:M:S
	dt_string = now.strftime("%Y-%m-%d %H:%M:%S")
	for purchase in eval(purchase_name):
		current_status = purchase['status']
		current_purchase_name = purchase['name']
		get_next_state = frappe.db.sql("""select next_state from `tabWorkflow Transition` where state = %s and action = %s""",(current_status,action),as_dict =1)
		next_state = get_next_state[0].next_state
		get_doc_status = frappe.db.sql("""select doc_status from `tabWorkflow Document State` where state = %s""",get_next_state[0].next_state, as_dict=1)
		doc_status = get_doc_status[0].doc_status
		if str(doc_status) == "0":
			flag = 1
			doc = frappe.get_doc("Purchase Order",current_purchase_name)

			update_statues = frappe.db.sql("""update `tabPurchase Order`  set workflow_state = %s , docstatus = %s, modified = %s where name = %s """, (next_state,doc_status,dt_string,current_purchase_name))
			update_status_items = frappe.db.sql("""update `tabPurchase Order Item`  set docstatus = %s , modified = %s where parent = %s """, (doc_status,dt_string,current_purchase_name))
		elif str(doc_status) == "1":
			flag = 1
			status = "To Receive and Bill"
			update_status = frappe.db.sql("""update `tabPurchase Order`  set workflow_state = %s , docstatus = %s, status = %s, modified = %s where name = %s """, (next_state,doc_status,status,dt_string,current_purchase_name))
			update_status_items = frappe.db.sql("""update `tabPurchase Order Item`  set docstatus = %s , modified = %s where parent = %s """, (doc_status,dt_string,current_purchase_name))
			#doc = frappe.get_doc("Purchase Order",current_purchase_name)
			#doc.update()
	return flag
def get_items_detials(name):
	items = frappe.db.sql("""select * from `tabPurchase Order Item` where parent = %s""",name,as_dict=1)
	return items
def get_tax_detials(name):
	taxes = frappe.db.sql("""select * from `tabPurchase Taxes and Charges` where parent = %s""",name, as_dict=1)
	return taxes
def get_payment_details(name):
	payment = frappe.db.sql("""select * from `tabPayment Schedule` where parent = %s""",name,as_dict=1)
	return payment
